### PR TITLE
Cleanup and promisify acceptance tests. Run tests for all backends.

### PR DESCRIPTION
**Pull request by gh2k**
### Description

This does decaf cleanup for the acceptance tests, and refactors them to run multiple configurations so that they can be run against different backends.

The idea here is that we can modify non-default backends and still test them, enabling us to add more backends in the future.

#### Related Issues / PRs

Initial decaf was #60 
Targeted for #67 

### Review

This required a fairly major refactor of `FilestoreApp` in order to support being torn down and brought up again, including a method to `disrequire` all files from the app directory. - It seems it isn't enough to just disrequire `app.js`, unfortunately.

Renamed the main file from `SendingFileTest` to `FilestoreTests` as this seemed more appropriate. The tests here test all of the functionality, not just sending files. Git doesn't seem to have tracked the rename - I think the files were too different.

### CODE REVIEW

File: test/acceptance/js/FilestoreApp.js [hunk 1]
```diff
@@ -1,109 +1,112 @@
...
+        throw err
+      }
+    }
+
+    this.initing = false
+  }
+
+  async waitForInit() {
+    while (this.initing) {
+      await new Promise(resolve => setTimeout(resolve, 1000))
```
Comment (emcsween) [inline test/acceptance/js/FilestoreApp.js]: setTimeout can be promisified via the magic of [util.promisify.custom](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_custom_promisified_functions)
```
const sleep = promisify(setTimeout)
await sleep(1000)
```

File: test/acceptance/js/FilestoreApp.js [hunk 2]
```diff
@@ -1,109 +1,112 @@
...
-        ? Settings.filestore.s3.endpoint
-        : undefined)
-    ) {
-      return callback()
+  }
+
+  async stop() {
+    if (this.server) {
+      await new Promise(resolve => {
+        this.server.close(resolve)
```
Comment (emcsween) [inline test/acceptance/js/FilestoreApp.js]: There's no error handling if something goes wrong closing the server. How about:
```
const closeServer = promisify(this.server.close).bind(this.server)
try {
  await closeServer()
} finally {
  delete this.server
}
```

File: test/acceptance/js/FilestoreApp.js [hunk 3]
```diff
@@ -1,109 +1,112 @@
...
-      (err, response) => {
-        console.log(
-          err,
-          response != null ? response.statusCode : undefined,
-          tries
+    let s3Available = false
+
+    while (tries < S3_TRIES && !s3Available) {
+      try {
+        const response = await promisify(request.get)(
```
Comment (emcsween) [inline test/acceptance/js/FilestoreApp.js]: You could use request-promise-native here

File: test/acceptance/js/FilestoreApp.js [hunk 4]
```diff
@@ -1,109 +1,112 @@
...
-        ) {
-          return callback()
+        if ([200, 404].includes(response.statusCode)) {
+          s3Available = true
         }
-
-        if (tries === S3_TRIES) {
-          return callback('timed out waiting for S3')
+      } catch (err) {
+      } finally {
```
Comment (emcsween) [inline test/acceptance/js/FilestoreApp.js]: I would add a comment in the `catch` clause indicating that errors are all swallowed on purpose.

File: test/acceptance/js/FilestoreApp.js (hunk 1 above)
Comment (gh2k) [reply test/acceptance/js/FilestoreApp.js]: TIL. Thanks :-)

File: test/acceptance/js/FilestoreApp.js (hunk 2 above)
Comment (gh2k) [reply test/acceptance/js/FilestoreApp.js]: nice :+1:

**Comment by gh2k** [comment]
Good feedback, thanks for the check. :-)

**Pull request merged**